lizrd/core/distributed.py

```diff
@@ -0,0 +1,94 @@
+from typing import Optional
+
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp import MixedPrecision, CPUOffload
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.nn as nn
+import torch
+
+from lizrd.core.misc import Noop
+
+
+def wrap_module_in_fsdp(
+    enabled: bool,
+    module: nn.Module,
+    rank: Optional[int],
+    param_precision: torch.dtype,
+    cast_inputs: bool = False,
+    offload_params: bool = False,
+    print_model: bool = False,
+):
+    wrapped = FSDP(
+        module,
+        device_id=rank,
+        mixed_precision=MixedPrecision(
+            param_dtype=param_precision,
+            reduce_dtype=torch.float32,
+            cast_forward_inputs=cast_inputs,
+        ),
+        cpu_offload=CPUOffload(offload_params=offload_params),
+    )
+    if print_model:
+        print("------- MODEL AFTER WRAPPING IN FSDP -------")
+        print(wrapped)
+        print("--------------------------------------------")
+    return wrapped
+
+
+def wrap_in_fsdp(
```

Comment:
what is the name responsible for? there's `wrap_module_in_fsdp`. What is being wrapped here?

Reply:
This is the function that wraps in fsdp. wrap_module_in_fsdp is a helper function that wraps one module. The reason for splitting them is that if we want to cast outputs, we need to create two modules and wrap each one in their own fsdp, then wrap both in cast wrapper.

Reply:
OK - `wrap_module_in_fsdp` wraps a single module. What does `wrap_in_fsdp` wrap?

Reply:
`wrap_in_fsdp` only passes arguments straight to FSDP and prints its parameters.
I would rename it to "_create_and_print_fsdp_model" and move it as an internal function of wrap_in_fsdp